src/backend/distributed/executor/repartition_logic.c

```diff
@@ -0,0 +1,252 @@
+ *
+ * Copyright (c) Citus Data, Inc.
+ */
+
+#include "postgres.h"
+#include "access/hash.h"
+#include "distributed/hash_helpers.h"
+
+#include "distributed/directed_acylic_graph_execution_logic.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/adaptive_executor.h"
+#include "distributed/worker_manager.h"
+#include "distributed/multi_server_executor.h"
+#include "distributed/repartition_logic.h"
+#include "distributed/worker_transaction.h"
+#include "distributed/worker_manager.h"
+#include "distributed/transaction_management.h"
+#include "distributed/multi_task_tracker_executor.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/transmit.h"
+
+
+static void FillTaskGroups(List **allTasks, List **mergeTasks);
+static List * CreateTemporarySchemas(List *mergeTasks);
+static List * CreateJobIds(List *mergeTasks);
+static void CreateSchemasOnAllWorkers(char *createSchemasCommand);
+static char * GenerateCreateSchemasCommand(List *jobIds);
+static char * GenerateCommand(List *jobIds, char *templateCommand);
+static bool DoesJobIDExist(List *jobIds, uint64 jobId);
+static void SendCommandToAllWorkers(List *commandList);
+static char * GenerateDeleteJobsCommand(List *jobIds);
+static void RemoveTempJobDirs(List *jobIds);
+
+
+/*
+ * ExecuteDependedTasks executes all tasks except the top level tasks
+ * in order from the task tree. At a time, it can execute different tasks from
+ * different jobs.
+ */
+void
+ExecuteDependedTasks(List *topLevelTasks)
+{
+	List *mergeTasks = NIL;
+
+
+	EnsureNoModificationsHaveBeenDone();
+
+	List *allTasks = TaskAndExecutionList(topLevelTasks);
+
+	FillTaskGroups(&allTasks, &mergeTasks);
+
+	List *jobIds = CreateTemporarySchemas(mergeTasks);
+
+	ExecuteTasksInDependencyOrder(allTasks, topLevelTasks);
+
+	RemoveTempJobDirs(jobIds);
+}
+
+
+/*
+ * FillTaskGroups iterates all tasks and creates a group for mergeTasks.
+ */
+static void
+FillTaskGroups(List **allTasks, List **mergeTasks)
+{
+	ListCell *taskCell = NULL;
+
+	foreach(taskCell, *allTasks)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+
+		if (task->taskType == MERGE_TASK)
+		{
+			*mergeTasks = lappend(*mergeTasks, task);
+		}
+	}
+}
+
+
+/*
+ * CreateTemporarySchemas creates the necessary schemas that will be used
+ * later in each worker. Single transaction is used to create the schemas.
+ */
+static List *
+CreateTemporarySchemas(List *mergeTasks)
+{
+	List *jobIds = CreateJobIds(mergeTasks);
+	char *createSchemasCommand = GenerateCreateSchemasCommand(jobIds);
+	CreateSchemasOnAllWorkers(createSchemasCommand);
+	return jobIds;
+}
+
+
+/*
+ * CreateJobIds returns a list of unique job ids that will be used
+ * in mergeTasks.
+ */
+static List *
+CreateJobIds(List *mergeTasks)
+{
+	ListCell *taskCell = NULL;
+	List *jobIds = NIL;
+
+	foreach(taskCell, mergeTasks)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		if (!DoesJobIDExist(jobIds, task->jobId))
```

Comment:
jobIds = list_append_unique_int(jobIds, task->jobIds);

Reply:
ah, I guess the problem is that it's int64

Reply:
I guess we cannot use that here because jobIds are uint64 not int.

Reply:
I guess it would be nice if Github had a feature like "there is a new comment" if I am sending a comment and a new one is added since the last time I refreshed the page.

Reply:
Well I have added a utility function for uint64, might be useful in future.